### ISSUE TITLE
unique validation doesn't work

### ISSUE THREAD

**Issue by angel1st** [balderdashy/sails#2823]
Hi there,
I have following User model:

    module.exports = _.merge(_.cloneDeep(require('../base/Model')), {
	attributes: {
		username: {
		  type: 'string',
		  unique: true
		},
		email: {
		  type: 'email',
		  unique: true
		},
		firstName: {
		  type: 'string',
		  required: true
		},
		lastName: {
		  type: 'string',
		  required: true
		},
		admin: {
		  type: 'boolean',
		  defaultsTo: false
		},

		// Below is all specification for relations to another models

		// Passport configurations
		passports: {
		  collection: 'Passport',
		  via: 'user'
		},
		// Login objects that are attached to user
		logins: {
		  collection: 'UserLogin',
		  via: 'user'
		},
		requestLogs: {
		  collection: 'RequestLog',
		  via: 'user'
		}
	}
    });

According to the above model declaration, I would expect when attempting to create a new user with already existing `username` and / or `email` to have exception returned. Instead I am able to successfully create as many users with identical usernames and passwords as I would like. 
I am using MongoDB and  sails-mongo, if that matters.

Best Regards,
Angel

**Comment by gilesbutler** [comment]
+1 @angel1st - I have exactly the same problem and it's preventing me from going live.

**Comment by angel1st** [comment]
@gilesbutler - I could suggest you workaround solution I use by myself - you could define unique index within the database, so you could implement this constraint on DB level. Of course this is not the best solution, but it works at least.
BTW - I am quite confusing that during the last couple of weeks feedback from the Sails supporting team is almost missing... hope this is just and accident, and they will back on track again soon.

**Comment by gilesbutler** [comment]
Thanks @angel1st - I tried this but the user doesn't get back an errors that are friendly so its a very bad user experience as the page just reloads and they don't know what's happened, maybe that's just my setup though.

There is an [open issue](https://github.com/balderdashy/sails-mongo/issues/193) for this in the sails-mongo repo anyway, hopefully they are working on a fix.

I'm not sure about the feedback from the Sails support team, maybe @tjwebb can shed some light on that one. I noticed the *needs failing test* label, @tjwebb is there anything we can do to help?

**Comment by angel1st** [comment]
@gilesbutler - I would suggest another workaround - you can check for uniqueness before adding new record, using `Model.find()` or `Model.findOne()` and if you get result(s), you can process the exception in more user friendly manner... hope this helps.

**Comment by gilesbutler** [comment]
@angel1st thanks for that, I thought that too. Is it a big performance hit to be hitting the DB twice though, once to check and then a second time to add a new record if none exists?

**Comment by angel1st** [comment]
@gilesbutler - well it depends of DB size, record / document size, but in general since this is just a read request for specific DB column and since the lack of other alternatives, I would suggest you to go for it and re-work your code at later stage.

**Comment by gilesbutler** [comment]
Thanks @angel1st - good to know and seems like the best solution for now! Thanks for your help.

**Comment by Nitemaeric** [comment]
Just dropping in to check if this issue is still being worked on?

This issue has occurred with multiple versions of waterline and sails. This should be added to the test list.

**Comment by luisincrespo** [comment]
Same issue here!. Any news?

**Comment by praveenaj** [comment]
Guys, try this: 
http://stackoverflow.com/questions/22868420/how-do-i-handle-a-unique-field-in-sails
It works like a charm now!

**Comment by sailsbot** [comment]
Thanks for posting, @angel1st.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
